src/renderer/viz/expressions/ramp.js

```diff
@@ -135,6 +143,262 @@ export default class Ramp extends BaseExpression {
+     *   // ]
+     * });
+     *
+     * @example <caption>Get 4 samples for a linear color ramp (String)</caption>
+     * const viz = new carto.Viz(`
+     *   color: ramp(linear($numeric, 1, 100), PRISM)
+     * `);
+     *
+     * layer.on('loaded', () => {
+     *   const legend = layer.getViz().color.getLegend({
+     *       samples: 4
+     *   });
+     *
+     *   // legend = [
+     *   //   { name: 'numeric', values: [// rgba color, [0, 25 ]] },
+     *   //   { name: 'numeric', values: [// rgba color, [25, 50 ]] }
+     *   //   { name: 'numeric', values: [// rgba color, [50, 75 ]] }
+     *   //   { name: 'numeric', values: [// rgba color, [75, 100 ]] }
+     *   // ]
+     * });
+     * @memberof carto.expressions.Ramp
+     * @name getLegend
+     * @instance
+     * @api
+     */
+    getLegend (options) {
+        const config = Object.assign(DEFAULT_OPTIONS, options);
+
+        if (config.samples > MAX_SAMPLES) {
+            throw new Error(`The maximum number of samples for a legend is ${MAX_SAMPLES}`);
+        }
+
+        if (this.input.isA(Linear)) {
+            return this._getLegendLinear(this.input, config);
+        }
+
+        if (this.input.type === inputTypes.CATEGORY) {
+            return this.input.list
+                ? this._getLegendList(config)
+                : this._getLeyendCategories(config);
+        }
+    }
+
+    _getLegendLinear (input, config) {
+        const firstChild = input[input.childrenNames[0]];
+        return firstChild.isA(Property)
+            ? this._getLegendLinearProperty(firstChild, config)
+            : this._getLegendLinear(firstChild, config);
+    }
+
+    _getLegendLinearProperty (property, config) {
+        const name = property.name;
+        const feature = {};
+        const breakpoints = _getBreakpoints(this._metadata.properties[name], config.samples);
+
+        return breakpoints.map((breakpoint) => {
+            feature[name] = breakpoint[0];
+            const values = [ this.eval(feature), breakpoint ];
+
+            return { name, values };
+        });
+    }
+
+    _getLegendList () {
+        return this.input.list.elems
+            .map(this._getLegendItemValue.bind(this))
+            .filter(legend => legend.values !== null);
+    }
+
+    _getLeyendCategories (config) {
+        return this.input.categories
+            .map((category, index) => { return this._getLegendCategoryValue(category, index, config); })
+            .filter(legend => legend.values !== null);
+    }
+
+    _getLegendItemValue (category, index) {
+        const value = this._getRampValueByIndex(index);
+
+        return {
+            name: this._getCategoryName(category.expr),
+            values: value ? [ value ] : null
+        };
+    }
+
+    _getLegendCategoryValue (category, index, config) {
+        const value = this._getRampValueByIndex(index);
+
+        if (this.input.isA(Top) && this.input.numBuckets === index) {
```

Comment:
I think the problem here is more in general, not just with this line, but the line is a good example.

The problem I see is that instead of using the public API with `eval()` which I think it's powerful enough for most of the implementation here, we are using introspection and private APIs to go case by case. 

I know that `eval` may not be enough or may need to be improved in some cases, but I think that's the way to go since I would expect less complexity (less case by case managament) and less code.